feat(mme): AMF Support for default slice type

panyogesh opened the pull request:
Fixed :
  1. Support added for default slice type (sST/SD) in mconfig.
  2. Support added for PLMN SUPPORT in config file having
     plmn to slice mapping in ngap response.
  3. Fixed te configuration read issue for amf name.
  4. Removed a global varaible buf_plmn.
  5. Add unit test infra for ngap setup request handler.

Test:
  1. Unit testing.
  2. Call flow testing with simulators.
 
Scenarios Covered :
1. No SliceType / No Slice Differentiator
2. With Slice Type as 1
3. With Mismatch Slice Types
4. With Matching SD Type
5. With Mismatching SD Type

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

<!-- Enumerate changes you made and why you made them -->

## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

panyogesh committed:
feat(mme): AMF Support for default Slice Type

Fixed :
  1. Support added for default slice type (sST/SD) in mconfig.
  2. Support added for PLMN SUPPORT in config file having
     plmn to slice mapping in ngap response.
  3. Fixed te configuration read issue for amf name.
  4. Removed a global varaible buf_plmn.
  5. Add unit test infra for ngap setup request handler.

Test:
  1. Unit testing.
  2. Call flow testing with simulators.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

panyogesh commented:
<img width="960" alt="Ngap-Setup-Response-with-SD-Sep-15" src="https://user-images.githubusercontent.com/69527565/133315990-764cc528-3d0d-4923-a00e-eb6a9c0a1e80.png">

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c:
```diff
@@ -500,159 +534,144 @@ status_code_e ngap_amf_handle_ng_setup_request(
+  Ngap_PLMNSupportList_t* plmn_support_list = NULL;
 
-  pdu.present = Ngap_NGAP_PDU_PR_successfulOutcome;
+  amf_config_read_lock(&amf_config);
```
I don't see where you are unlocking this.

ardzoht commented on lte/protos/mconfig/mconfigs.proto:
```diff
@@ -356,6 +356,24 @@ message MME {
     SentryConfig sentry_config = 45;
     
     bool enable_converged_core = 46;
+
```
Thanks for updating these

ardzoht commented on lte/gateway/configs/gateway.mconfig:
```diff
@@ -39,7 +39,8 @@
       "congestionControlEnabled": true,
-      "enableConvergedCore": false
+      "enableConvergedCore": false,
+      "defaultSliceServiceType": "1",
```
this should not have a trailing comma, as this is the last entry of the JSON object
```suggestion
      "defaultSliceServiceType": "1"
```

ardzoht reviewed (changes requested):
let's remove the trailing comma on `defaultSliceServiceType` on the default gateway.mconfig, as this will break the JSON parsing

ulaskozat commented on lte/gateway/configs/templates/mme.conf.template:
```diff
@@ -352,4 +353,8 @@ NGAP :
     DEFAULT_DNS_SEC_IPV4_ADDRESS = "{{ ipv4_sec_dns }}";
+    AMF_NAME = "{{ amf_name }}";
+    PLMN_SUPPORT_LIST = (
+       { MCC="{{ mcc }}" ; MNC="{{ mnc }}"; DEFAULT_SLICE_SERVICE_TYPE="{{ default_slice_service_type }}" ;                         DEFAULT_SLICE_DIFFERENTIATOR="{{ default_slice_differentiator }}"; }
```
whys is there so many blank spaces before `DEFAULT_SLICE_DIFFERENTIATOR`?

panyogesh replied:
Comments addressed.

panyogesh committed:
feat(mme): Addressing review comments - 9048

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/tasks/ngap/ngap_amf_handlers.c:
```diff
@@ -500,159 +534,145 @@ status_code_e ngap_amf_handle_ng_setup_request(
-      SliceItem);  // adding slice item to slice list
+status_code_e ngap_generate_ng_setup_response(
+    ngap_state_t* state, gnb_description_t* gnb_association) {
+  int i;
```
legit github comment about unused variable.

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/c/core/oai/tasks/amf/amf_config.c:
```diff
@@ -662,7 +665,98 @@ int amf_config_parse_file(amf_config_t* config_pP) {
+          } else {
+            config_pP->plmn_support[i].s_nssai.sst =
+                NGAP_S_NSSAI_ST_DEFAULT_VALUE;
+          }
```
we should not do this here. Please add it into config initialization and here it can be overwritten if specified in the config.

ardzoht replied:
+1

panyogesh replied:
Comments addressed.

panyogesh committed:
feat(mme): Updated Review comments 9048

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

panyogesh commented:
<img width="960" alt="Default-NGAP-Setup-Request-Sep-16" src="https://user-images.githubusercontent.com/69527565/133617617-7826b603-8a07-4787-81db-5a3dd67ae558.png">
<img width="960" alt="Ngap-Setup-Request-with-SD-Sep-16" src="https://user-images.githubusercontent.com/69527565/133617621-26cda9bc-b60f-4305-a379-0911738a637b.png">
<img width="960" alt="Ngap-Setup-Response-with-SD-Sep-16" src="https://user-images.githubusercontent.com/69527565/133617625-ab51d597-e9b6-4cbd-b53a-da0ad974099e.png">

panyogesh commented:
[Test-Logs-Slice-SD-Support-Sep-16.txt](https://github.com/magma/magma/files/7177921/Test-Logs-Slice-SD-Support-Sep-16.txt)

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -282,8 +282,7 @@ def _get_default_slice_differentiator_type_config(service_mconfig: object) -> st
     """
-
-    if service_mconfig.default_slice_differentiator is not None:
+    if len(service_mconfig.default_slice_differentiator):
```
what is the reason for this change?

panyogesh replied:
With GRPC/Protobuf an empty string is "" (len = 0)... Tested with following program to confirm                                
    if service_mconfig.default_slice_differentiator is not None:
        os.write(fd, str.encode (" ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -"))
        os.write(fd, str.encode(str(len(service_mconfig.default_slice_differentiator))))

Output : ** SLICE DIFFERNTIATOR IS NOT NONE ** : len -0

Hence in mme.conf the default value "" and not the Default Value return by generate_oai_config script..

I guess we need to change other string variables on similar lines.

ulaskozat replied:
if these things are undefined on the orc8r side, I think the generated gateway.mconfig will not have the fields at all.  @ardzoht what do you say?

ardzoht replied:
The only case where you should expect an empty string from generated mconfigs is if the cloud config specifically returns that, but that should not be the case, as the default value for string types on the cloud configs is `"string"`. 

And if the config is not specified (as Ulas points out), there won't be any entry on the config at all. So `if service_mconfig.default_slice_differentiator is not None` is the correct handler for this. @panyogesh can you update it?

panyogesh replied:
[Length-Experiment-Output.txt](https://github.com/magma/magma/files/7180279/Length-Experiment-Output.txt)

Attaching the output of experiment.  Printing default value in both cases.  EXPECTED DEFAULT VALUE : "0xffffff"

ulaskozat replied:
@panyogesh is orc8r connected to your set up? If service_mconfig.default_slice_differentiator is None, then len will fail.

ardzoht replied:
There are a couple of things to note here:

1. Currently there is no `defaultSliceDifferentiator` entry on the gateway.mconfig, so there won't be a parsed entry on the `service_mconfig` python object
2. Checking for `service_mconfig.default_slice_differentiator` existance should be enough, as Ulas says the call on `len()` will complain if there is no entry on there, which is this exact case.

So my comment still applies, I think we should just check for the entry existance `if service_mconfig.default_slice_differentiator is not None` to handle this entry. And it would be a good idea to include an entry for this on the `gateway.mconfig` default json file.

ardzoht replied:
And just to note, currently there is no Orc8r configs to build from for these entries, if the plan is to also add the configs to the cloud, `magma/lte/cloud/go/services/lte/obsidian/models/swagger.v1.yml` should be updated to include these configs (either on gateway or network EPC configs?), this can be done on a separate PR but just wanted to bring the awareness for this. cc @ulaskozat

ulaskozat commented:
@rdefosse is Magma-OAI-Jenkins hanging?

ulaskozat reviewed (approved):
LGTM with a nit comment

ardzoht reviewed (approved):
lgtm

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if len(service_mconfig.default_slice_service_type):
+        return service_mconfig.default_slice_service_type
+
+    return DEFAULT_NGAP_S_NSSAI_SST
```
instead use `return service_mconfig.default_slice_service_type or DEFAULT_NGAP_S_NSSAI_SST`

ulaskozat replied:
cc @ardzoht

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if len(service_mconfig.default_slice_differentiator):
+        return service_mconfig.default_slice_differentiator
+
+    return DEFAULT_NGAP_S_NSSAI_SD
```
instead use `return service_mconfig.default_slice_differentiator or DEFAULT_NGAP_S_NSSAI_SD`

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if service_mconfig.amf_name is not None:
+        return service_mconfig.amf_name
+
+    return DEFAULT_NGAP_AMF_NAME
```
instead use `return service_mconfig.amf_name or DEFAULT_NGAP_AMF_NAME`

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if service_mconfig.amf_region_id is not None:
+        return service_mconfig.amf_region_id
+
+    return DEFAULT_NGAP_AMF_REGION_ID
```
instead use `return service_mconfig.amf_region_id or DEFAULT_NGAP_AMF_REGION_ID`

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if service_mconfig.amf_set_id is not None:
+        return service_mconfig.amf_set_id
+
+    return DEFAULT_NGAP_SET_ID
```
instead use `return service_mconfig.amf_set_id or DEFAULT_NGAP_SET_ID`

panyogesh replied:
Comments addressed.

ulaskozat commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -251,6 +257,101 @@ def _get_converged_core_config(service_mconfig: object) -> bool:
+    if service_mconfig.amf_pointer is not None:
+        return service_mconfig.amf_pointer
+
+    return DEFAULT_NGAP_AMF_POINTER
```
instead use `return service_mconfig.amf_pointer or DEFAULT_NGAP_AMF_POINTER`

panyogesh replied:
Comments addressed.

ardzoht commented on lte/gateway/configs/gateway.mconfig:
```diff
@@ -39,7 +39,8 @@
       "natEnabled": true,
       "congestionControlEnabled": true,
-      "enableConvergedCore": false
+      "enableConvergedCore": false,
```
add entry for `defaultSliceDifferentiator` here as well

panyogesh replied:
Comments addressed.

panyogesh committed:
feat(mme): Addressed review comments

1. Fixed the generate_oai file changes
2. Added default slice descriptor
3. Fixed formatting issues.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

ulaskozat commented:
@panyogesh have you regenerated the pb files? insync-checkin is failing.

ulaskozat commented on lte/gateway/configs/gateway.mconfig:
```diff
@@ -39,7 +39,9 @@
-      "enableConvergedCore": false
+      "enableConvergedCore": false,
+      "defaultSliceServiceType": "1",
+      "defaultSliceDifferentiator" : "0x0"
```
why is this default different than the defaults defined in other parts?

panyogesh replied:
Comments addressed. Changed it to 0xffffff

panyogesh committed:
feat(mme) : Proto file and default sd value change

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

panyogesh commented:
<img width="960" alt="Configured-SD-NGAP-Setup-Request-Sep-17" src="https://user-images.githubusercontent.com/69527565/133823704-dd13823c-57c3-47d6-98e9-3aff9963b9fc.png">
<img width="960" alt="Default-NGAP-Setup-Request-Sep-17" src="https://user-images.githubusercontent.com/69527565/133823716-36a25e94-c157-4ffb-bd78-8c31fb9ef098.png">

ardzoht commented on lte/gateway/python/scripts/generate_oai_config.py:
```diff
@@ -38,6 +38,12 @@
+DEFAULT_NGAP_AMF_NAME = "AMF_1"
+DEFAULT_NGAP_AMF_REGION_ID = "1"
+DEFAULT_NGAP_SET_ID = "1"
+DEFAULT_NGAP_AMF_POINTER = "1"
```
this value for amf pointer is 1 but on the gateway.mconfig is set to 3, let's set them both to the expected default value. cc @ulaskozat

panyogesh replied:
Comments addressed. The default value of pointer should be 0.

panyogesh commented:
<img width="960" alt="Default-NGAP-Setup-Request-Sep-17-1" src="https://user-images.githubusercontent.com/69527565/133833318-ec4b6c94-0712-46ae-b192-98fdeb286cb8.png">
